src/DocumentationViewer.js

```diff
@@ -0,0 +1,219 @@
+  application
+  }) {
+  const builder = Gtk.Builder.new_from_resource(resource);
+  const window = builder.get_object("documentation_viewer");
+
+  const container = builder.get_object("webview_container");
+  const webview = new WebKit.WebView();
+  const listbox = builder.get_object("listbox");
+
+  const sidebar_button = builder.get_object("sidebar_button");
+  const search_button = builder.get_object("search_button");
+  const search_entry = builder.get_object("search_entry");
+  const search_bar = builder.get_object("search_bar");
+  const back = builder.get_object("back");
+  const forward = builder.get_object("forward");
+
+  const base_path = "/app/share/doc";
+
+  webview.load_uri(`file://${base_path}/gtk4/index.html`);
+  container.child = webview;
+  let loaded = false;
+
+  webview.connect("load-changed", (view, load_event) => {
+    if (load_event === WebKit.LoadEvent.FINISHED) {
+      loaded = true;
+      sidebar_button.active = false;
+    } else {
+      loaded = false;
+    }
+  });
+
+  back.connect("clicked", () => {
+    webview.go_back();
+  });
+
+  forward.connect("clicked", () => {
+    webview.go_forward();
+  });
+
+  sidebar_button.connect("toggled", () => {
+    if (loaded) {
+      if (sidebar_button.active) disableDocSidebar(webview);
+      else enableDocSidebar(webview);
+    }
+  });
+
+  sidebar_button.connect("toggled", () => {
+    if (loaded) {
+      if (sidebar_button.active) disableDocSidebar(webview);
+      else enableDocSidebar(webview);
+    }
+  });
+
+  search_entry.connect("search-changed", () => {
+    listbox.invalidate_filter();
+  });
+
+  search_bar.bind_property(
+    "search-mode-enabled",
+    search_button,
+    "active",
+    GObject.BindingFlags.BIDIRECTIONAL,
+  );
+
+  search_button.connect("toggled", () => {
+    if (search_button.active) sidebar_button.active = true;
+  });
+
+  listbox.set_sort_func(sort);
+  listbox.invalidate_sort();
+  listbox.set_filter_func(filter);
+
+  getDocs(base_path)
+    .then((docs) => {
+      for (const doc in docs) {
+        const row = new Adw.ActionRow({
+          title: doc,
+        });
+        row.add_suffix(new Gtk.Image({ icon_name: "go-next-symbolic" }));
+        listbox.append(row);
+      }
+      listbox.connect("row-selected", (_, row) => {
+        webview.load_uri(docs[row.title]);
+      });
+      return docs;
+    })
+    .catch(logError)
+
+
+
+  function sort(row1, row2) {
+    return row1.title > row2.title;
+  }
+
+  function filter(row) {
+    const re = new RegExp(search_entry.text, "i");
+    return re.test(row.title);
+  }
+
+  const action_documentation = new Gio.SimpleAction({
+    name: "documentation",
+    parameter_type: null,
+  });
+  action_documentation.connect("activate", () => {
+    window.present();
+  });
+  application.add_action(action_documentation);
+  application.set_accels_for_action("app.documentation", ["<Control><Shift>D"]);
+}
+
+async function getDocs(base_path) {
+  try {
+    const docs = {};
```

Comment:
please turn docs into an array of objects instead

Reply:
I did not turn docs into an array of objects because I didn't want to iterate over it every time the user selects a doc, and check which doc was selected in the listbox, instead I could directly get the uri of the selected doc. Would the array make it inefficient or will it be fine?

Reply:
We don't need to iterate over the docs everytime see https://github.com/sonnyp/Workbench/pull/358#discussion_r1245883792